client/ring_buffer.go

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2022 Contributors to the Eclipse Foundation
+//
+// See the NOTICE file(s) distributed with this work for additional
+// information regarding copyright ownership.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0
+//
+// SPDX-License-Identifier: EPL-2.0
+
+package client
+
+import "github.com/eclipse-kanto/file-upload/logger"
+
+// ringBuffer is a circular buffer of interface{} elements with fixed capacity.
+// Oldest elements are overwritten, when there is no more space in the buffer.
+type ringBuffer struct {
+	elements []interface{}
+	start    int
+	end      int
+}
+
+// newRingBuffer creates a new ringBuffer with the given size.
+func newRingBuffer(size int) *ringBuffer {
+	buf := &ringBuffer{}
+
+	buf.elements = make([]interface{}, size+1)
+
+	return buf
+}
+
+// empty returns true, if the buffer is empty.
+func (buf *ringBuffer) empty() bool {
+	return buf.start == buf.end
+}
+
+// get the element at the beginning of the buffer, log error if empty.
+func (buf *ringBuffer) get() interface{} {
+	if buf.empty() {
+		logger.Error("cannot get element from empty buffer")
```

Comment:
Just log error doesn't work. If the ring buffer is used as a pure data structure, it should not log error but return value/error.